## Linear Regression

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sklearn
from numpy import array
from numpy import argmax
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, KFold
from sklearn import datasets, linear_model, metrics 
from sklearn.cross_validation import cross_val_score
import pickle

/home/atibhi/Desktop/mlhack/myenv/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/atibhi/Desktop/mlhack/myenv/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/atibhi/Desktop/mlhack/myenv/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/atibhi/Desktop/mlhack/myenv/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/atibhi/Desktop/mlhack/myenv/lib/python3.5/importlib/_bootstrap.py:222: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args

In [2]:
pd.options.mode.chained_assignment = None
train_data = pd.read_csv("training/training_data.csv")
train_data = train_data.drop(['Position'],axis=1)

In [3]:
X = train_data 
y = pd.read_csv("ground_truth/overall.csv") 

In [4]:
X.shape

(3147, 47)

In [5]:
y.shape

(3147, 1)

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, random_state=42) 

In [7]:
reg = linear_model.LinearRegression() 

In [ ]:
model = reg.fit(X_train, y_train) 

In [ ]:
pickle.dump(reg,open("LR_Pickles/linear_regression_goalie.p","wb"))

In [8]:
scores = cross_val_score(model,X_train,y_train,cv=5)
print(scores)

[0.72607626 0.79604779 0.78228138 0.76212918 0.76964534]


In [9]:
# regression coefficients 
print('Coefficients: \n', reg.coef_) 

Coefficients: 
 [[ 0.27635001  0.09794067 -0.01363103  0.21251474  0.04759819 -0.00597777
  -0.0160269   0.00223071  0.11813603  0.10092852  0.06012835 -0.05018245
   0.01717846 -0.06107797 -0.06154904  0.1201743   0.00451206  0.15699408
   0.01478658  0.10062327  0.21963664  0.00647186  0.58666418  0.04792763
  -0.04361057  0.01401454 -0.08028596 -0.02916311 -0.06094243  0.06736888
   0.05192033 -0.05350229  0.99974924 -0.23365423  0.01109688  0.65183993
   0.22806031  0.55043642 -0.70846122 -2.01806074 -0.57788532  0.89451162
   0.08030044 -1.34576434  1.24019049  0.12482611  1.53184622]]


In [10]:
# variance score: 1 means perfect prediction 
print('Variance score: {}'.format(reg.score(X_test, y_test))) 

Variance score: 0.768163589625066


## Results

For Linear Regression, accuracies of ratings of players are :

- **Goalkeeper** : 0.9447227122342368, **Cross-val** = [0.69046458 0.86651207 0.90495512 0.93145542 0.92278509]
- **Forward** :  0.8617509319866774, **Cross-val** = [0.84284416 0.88360999 0.80799645 0.86456835 0.84842202]
- **Defender** : 0.803729638101549, **Cross-val** = [0.80717527 0.75942803 0.84954936 0.71503158 0.79869502]
- **Midfielder** : 0.8395039300810384, **Cross-val** = [0.83831646 0.84487003 0.83584216 0.82997279 0.86120531]
- **Overall**: 0.768163589625066, **Cross-val** = [0.72607626 0.79604779 0.78228138 0.76212918 0.76964534]